pkg/agent/v1alpha1/server.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"net"
+	"os"
+	"sync"
+
+	"github.com/golang/glog"
+	"google.golang.org/grpc"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+)
+
+// TD: @amitd
+// This entire code needs to be structured & designed.
+// It is far far away from SOLID principles !!!!
+
+// NonBlockingGRPCServer defines Non blocking GRPC server interfaces
+type NonBlockingGRPCServer interface {
+	// Start services at the endpoint
+	Start(endpoint string, ids csi.IdentityServer, cs csi.ControllerServer, ns csi.NodeServer)
+
+	// Waits for the service to stop
+	Wait()
+
+	// Stops the service gracefully
+	Stop()
+
+	// Stops the service forcefully
+	ForceStop()
+}
+
+// NewNonBlockingGRPCServer returns a new instance of NonBlockingGRPCServer
+func NewNonBlockingGRPCServer() NonBlockingGRPCServer {
+	return &nonBlockingGRPCServer{}
+}
+
+// NonBlocking server
+type nonBlockingGRPCServer struct {
+	wg     sync.WaitGroup
+	server *grpc.Server
+}
+
+// Start grpc server for serving CSI endpoints
+func (s *nonBlockingGRPCServer) Start(endpoint string, ids csi.IdentityServer, cs csi.ControllerServer, ns csi.NodeServer) {
+
+	s.wg.Add(1)
+
+	go s.serve(endpoint, ids, cs, ns)
+
+	return
+}
+
+// Wait for the service to stop
+func (s *nonBlockingGRPCServer) Wait() {
+	s.wg.Wait()
+}
+
+// Stop the service forcefully
+func (s *nonBlockingGRPCServer) Stop() {
+	s.server.GracefulStop()
+}
+
+// ForceStop the service
+func (s *nonBlockingGRPCServer) ForceStop() {
+	s.server.Stop()
+}
+
+// serve starts serving requests at the provided endpoint based on the type of
+// plugin. In this function all the csi related interfaces are provided by
+// container-storage-interface
+func (s *nonBlockingGRPCServer) serve(endpoint string, ids csi.IdentityServer, cs csi.ControllerServer, ns csi.NodeServer) {
```

Comment:
can we move the arguments to fields of nonBlockingGRPCStruct

Reply:
yeah